Fix typings for Mesh class

r-Larch opened the pull request:
Before this change the `material` property of the Mesh class was allways of the union type `Material | Material[]`
```typescript
new Mesh(geo, ...).material // -> showed the union type: Material | Material[]
// this resulted in:
new Mesh(geo, new MeshBasicMaterial(...)).material.clone() // -> Property 'clone' does not exist on type 'Material | Material[]'.
```

To fix this I added an implicit recognizable generic type `TMaterial` which can be some type which extends `Material` or `Material[]`
and to cover the case where the optional parameter `material` is not given I let `TMaterial` default to `MeshBasicMaterial` which
is the case in code see: https://github.com/mrdoob/three.js/blob/dev/src/objects/Mesh.js#L50

Now the `mesh.material` property shows the correct type.
```typescript
new Mesh(geo, new MeshBasicMaterial(...)).material // -> shows type MeshBasicMaterial
new Mesh(geo, [new MeshBasicMaterial(...)]).material // -> shows type MeshBasicMaterial[]
new Mesh(geo).material // -> shows type MeshBasicMaterial because the parameter defaults to that in the constructor: https://github.com/mrdoob/three.js/blob/dev/src/objects/Mesh.js#L50
```

r-Larch committed:
Fix typings for Mesh class

Before this change the `material` property of the Mesh class was allways of the union type `Material | Material[]`
```typescript
new Mesh(geo, ...).material // -> showed the union type: Material | Material[]
// this resulted in:
new Mesh(geo, new MeshBasicMaterial(...)).material.clone() // -> Property 'clone' does not exist on type 'Material | Material[]'.
```

To fix this I added a implicit recognisable generic type `TMaterial` witch can be some type witch extends `Material` or `Material[]`
and to cover the case where the optional parameter `material` is not given I let `TMaterial` default to `MeshBasicMaterial` witch 
is the case in code see: https://github.com/mrdoob/three.js/blob/dev/src/objects/Mesh.js#L50

Now the `mesh.material` property shows the correct type.
```typescript
new Mesh(geo, new MeshBasicMaterial(...)).material // -> shows type MeshBasicMaterial
new Mesh(geo, [new MeshBasicMaterial(...)]).material // -> shows type MeshBasicMaterial[]
new Mesh(geo).material // -> shows type MeshBasicMaterial because the parameter defaults to that in the constructor: https://github.com/mrdoob/three.js/blob/dev/src/objects/Mesh.js#L50
```

r-Larch committed:
fixes typings for `examples\jsm\objects\Fire.d.ts` witch extends Mesh<TMaterial>

Mugen87 commented on examples/jsm/objects/Fire.d.ts:
```diff
@@ -34,7 +34,7 @@ export interface FireOptions {
 }
 
-export class Fire extends Mesh {
+export class Fire extends Mesh<ShaderMaterial> {
```
There are some more TS files that define classes which extend from `Mesh` (e.g. `Lensflare`). Is it necessary to change these definitions in the same way? I mean adding `<some material>`.

r-Larch replied:
In case of [Lensflare](https://github.com/mrdoob/three.js/blob/e8145dcaa83f57da82a3d3d413c89e34b9f01571/examples/jsm/objects/Lensflare.d.ts) it isn't because it does not override the `material` property with some different type.

In other cases, where some type extends `Mesh` and overrides the `material` property, it is necessary to change the definition, because doing so ensures exact typings all the way through.

An alternate approach would be to let the generic `TMaterial` of the `Mesh` class default to `Material` (which seems to be the base of all materials) then it shouldn't be necessary to change `Fire.d.ts`, but with the consequences of the `material` property in `Fire` class not being exact typed.

r-Larch replied:
# The Alternate approach I mentioned earlier

> An alternate approach would be to let the generic TMaterial of the Mesh class default to 
> Material then it shouldn't be necessary to change Fire.d.ts, but with the consequences of the material property in Fire class not being exactly typed.

Using this approach shouldn't break any existing code by making some compromises.

So, let the generic `TMaterial` of the `Mesh` class default to `Material`

**Mesh.d.ts**
```diff
- export class Mesh<TMaterial extends Material | Material[] = MeshBasicMaterial> extends Object3D {
+ export class Mesh<TMaterial extends Material | Material[] = Material> extends Object3D {
```
**Filre.d.ts** stays as it allways was.

## Using it
```js
new Mesh(geo).material // -> shows type Material (in reality it is MeshBasicMaterial)
// ... all other behave like before:
new Mesh(geo, new MeshBasicMaterial(...)).material // -> shows type MeshBasicMaterial
new Mesh(geo, [new MeshBasicMaterial(...)]).material // -> shows type MeshBasicMaterial[]

const mesh = new Mesh(geo, [
   new MeshBasicMaterial(), 
   new MeshNormalMaterial(), 
   new MeshLambertMaterial(),
   new MeshNormalMaterial(),
]);
mesh.material; // shows -> (MeshBasicMaterial | MeshNormalMaterial | MeshLambertMaterial)[]
```

## Extending it
### case 1:

```js
class Lensflare extends Mesh {
   // the property `material` stays unchanged.
}
new Lensflare().material // shows -> `Material` (in reality it is MeshBasicMaterial)
```

### case 2:
```js
class Fire extends Mesh {
// override the material prop with different type:
   material: ShaderMaterial;
}
new Fire().material 
// should "normally" show -> `Material` (in reality it is ShaderMaterial)
// note: vscode is able to detect the real type here and shows `ShaderMaterial`
```

### case 3:
```js
// adding the type as generic argument on the base class
class FireWithGenericMeshClass extends Mesh<ShaderMaterial> {
   material: ShaderMaterial;
}
new FireWithGenericMeshClass().material // shows -> ShaderMaterial
```

### case 4:
```js
class Fire1 extends Mesh {
    constructor(/*...*/){
        super(/*...*/);
        
        // set the material prop to some different type:
        this.material = new ShaderMaterial();
    }
 }
 new Fire1().material // shows -> `Material` (in reality it is ShaderMaterial)
```

## Type checking
```js
class Fire2 extends Mesh<ShaderMaterial> {
 }
 new Fire2().material // shows -> ShaderMaterial

 // assiging the wrong type shows some type error!
 new Fire2().material = new MeshNormalMaterial(); // ❌
```

```js
var mesh = new Mesh(geo, [new MeshNormalMaterial(), new ShaderMaterial()]);
// assiging the wrong type shows some type error!
mesh.material = new MeshNormalMaterial(); // ❌
```

```js
{
    const myMaterial = [new MeshNormalMaterial(), new ShaderMaterial()];
    const mesh = new Mesh(geo, myMaterial);
    // pushing a wrong type to the array shows some type error!
    mesh.material.push(new MeshBasicMaterial()); // ❌
}
{
    // the type error can be fixed by specifing a common base type.
    const myMaterial: Material[] = [new MeshNormalMaterial(), new ShaderMaterial()];
    const mesh = new Mesh(geo, myMaterial);
    mesh.material.push(new MeshBasicMaterial()); // ✔
}
```

mrdoob commented:
```js
new Mesh(geo, [new MeshBasicMaterial(...)]).material // -> shows type MeshBasicMaterial[]
```

That looks weird to me..? What would the type be when doing this?

```js
new Mesh(geo, [new MeshBasicMaterial(), new MeshNormalMaterial()]).material
```

r-Larch commented:
> ...
> 
> That looks weird to me..? What would the type be when doing this?
> 
> ```js
> new Mesh(geo, [new MeshBasicMaterial(), new MeshNormalMaterial()]).material
> ```

It automaticly detects and shows an `Array` of the union type `MeshBasicMaterial | MeshNormalMaterial` see:
```typescript
new Mesh(geo, [new MeshBasicMaterial(), new MeshNormalMaterial()]).material 
// shows -> (MeshBasicMaterial | MeshNormalMaterial)[]

// Adding more different types just adds them to the union type 
const mesh = new Mesh(geo, [
   new MeshBasicMaterial(), 
   new MeshNormalMaterial(), 
   new MeshLambertMaterial(),
   new MeshNormalMaterial(), // Note this is hear twise, but not in the resulting union type!
]);
mesh.material; // shows -> (MeshBasicMaterial | MeshNormalMaterial | MeshLambertMaterial)[]
```

So any combination should work well.

Mugen87 commented:
Unfortunately, I'm not that familiar with TS so I'm not feeling confident to review this change. Maybe someone else with more TS experience can evaluate this PR.

Mugen87 commented:
@martinRenou What do you think about this PR? 😇

martinRenou commented:
I don't think I can call myself a TypeScript expert :P

But one thing I notice is I think it would make sense to do the same for `Geometry` in this PR. As a `Mesh` is really defined by a Geometry and a Material.
Other than that it looks good to me! It is a really nice addition and people will definitely benefit from it.

I guess it's backward-incompatible change though? @r-Larch if someone made a class inheriting from `Mesh` in TypeScript, they would need to update their code for it to compile, right?

r-Larch committed:
Merge branch 'dev' into patch-1

r-Larch commented:
We can make it backward-compatible if we let the generic-type `TMaterial` default to `Material` as I described in this [alternate approach](https://github.com/mrdoob/three.js/pull/17564#discussion_r328829817).
And the same could be done with `Mesh.geometry`.
(I think this is the preferred solution)

Required changes:
```typescript
...
export class Mesh<
	TMaterial extends Material | Material[] = Material,
	TGeometry extends Geometry | BufferGeometry = Geometry | BufferGeometry
	> extends Object3D {

	constructor(
		geometry?: TGeometry,
		material?: TMaterial
	);

	geometry: TGeometry;
	material: TMaterial;
        ...
```

Allows to do:
```typescript
const mesh = new Mesh(new BoxGeometry(), new ShaderMaterial()); 
mesh.geometry; // shows: BoxGeometry
mesh.material; // shows: ShaderMaterial
```

Extending:
```typescript
// implicit recognition of the types:
class TestMesh extends Mesh {
	material = new MeshNormalMaterial();
	geometry = new BoxGeometry();
}

new TestMesh().material // shows: MeshNormalMaterial
new TestMesh().geometry // shows: BoxGeometry
```

```typescript
// defining the types explizit:
class TestMesh extends Mesh<MeshNormalMaterial, BoxGeometry> {
}

new TestMesh().material // shows: MeshNormalMaterial
new TestMesh().geometry // shows: BoxGeometry
```

r-Larch committed:
make typings for Mesh class more generic

Allows:
```typescript
const mesh = new Mesh(new BoxGeometry(), new ShaderMaterial()); 
mesh.geometry; // shows: BoxGeometry
mesh.material; // shows: ShaderMaterial
```

Mugen87 commented:
Closing in favor of #19299.
